src/python/pants/backend/jvm/subsystems/graal.py

```diff
@@ -0,0 +1,182 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+import os
+import shutil
+
+from pants.backend.native.config.environment import Platform
+from pants.base.build_environment import get_pants_cachedir
+from pants.base.hash_utils import hash_file, stable_json_sha1
+from pants.base.workunit import WorkUnitLabel
+from pants.binaries.binary_tool import NativeTool
+from pants.binaries.binary_util import BinaryToolUrlGenerator
+from pants.option.compiler_option_sets_mixin import CompilerOptionSetsMixin
+from pants.util.contextutil import temporary_dir
+from pants.util.dirutil import safe_mkdir_for
+from pants.util.memo import memoized_method, memoized_property
+from pants.util.meta import classproperty
+from pants.util.process_handler import subprocess
+from pants.util.strutil import safe_shlex_join
+
+
+logger = logging.getLogger(__name__)
+
+
+class GraalCEUrlGenerator(BinaryToolUrlGenerator):
+
+  _DIST_URL_FMT = 'https://github.com/oracle/graal/releases/download/vm-{version}/{base}'
+  _ARCHIVE_BASE_FMT = 'graalvm-ce-{version}-{system_id}-amd64.tar.gz'
+  _SYSTEM_ID = {
+    'mac': 'macos',
+    'linux': 'linux',
+  }
+
+  def generate_urls(self, version, host_platform):
+    system_id = self._SYSTEM_ID[host_platform.os_name]
+    archive_basename = self._ARCHIVE_BASE_FMT.format(version=version, system_id=system_id)
+    return [self._DIST_URL_FMT.format(version=version, base=archive_basename)]
+
+
+class GraalCE(NativeTool, CompilerOptionSetsMixin):
+
+  options_scope = 'graal'
+  default_version = '1.0.0-rc15'
+  archive_type = 'tgz'
+
+  def get_external_url_generator(self):
+    return GraalCEUrlGenerator()
+
+  @classproperty
+  def get_compiler_option_sets_enabled_default_value(cls):
+    return {
+      'image-build-debug': [
+        '--verbose',
+        '--no-server',
+        '-H:+ReportExceptionStackTraces',
+      ],
+      'release': ['-O9'],
+    }
+
+  @classproperty
+  def get_compiler_option_sets_disabled_default_value(cls):
+    return {
+      'release': ['-O0'],
+      # If 'strict' is selected as an option set, the native image will build without these
+      # options. This isn't expected to be used, mostly using the option sets as a method of
```

Comment:
Having it in the compiler option sets disabled args means that the arguments here will be used on all command lines unless someone adds `strict` as an option set. Here it's used as a tool to organize what arguments are fed into the native image without hardcoding them.

Reply:
If I'm understanding this correctly, I think I would rather hard-code them, than having something appear as an option which is explicitly listed as an option you shouldn't enable because it's always enabled by default. Options that aren't options is a really confusing semantic.

Reply:
That makes sense. I'm thinking of moving these to a separate shlexed string list option than `--compiler-option-sets`, which was absolutely being abused here.

Reply:
Sounds good :) Thanks!